File: ParteC#/ProyectoProgramacion3/ProyectoProgramacion3Persistencia/DAO/reservas/ServicioDaoImpl.cs
using System.Data;
using System.Data.Common;
using ProyectoProgramacion3Model.Model.reservas;

namespace ProyectoProgramacion3Persistencia.DAO.reservas;

public class ServicioDaoImpl : DefaultBaseDao<Servicio>, IServicioDao
{
    protected override DbCommand ComandoCrear(DbConnection conn, Servicio modelo)
    {
        var cmd = CrearStoredProcedure(conn, "sp_InsertServicio");
        SetCamposServicio(cmd, modelo);
        CrearParametroSalida(cmd, "_id_generado", DbType.Int32);
        return cmd;
    }

    protected override DbCommand ComandoActualizar(DbConnection conn, Servicio modelo)
    {
        var cmd = CrearStoredProcedure(conn, "sp_UpdateServicio");
        CrearParametro(cmd, "p_id_servicio", modelo.idServicio);
        SetCamposServicio(cmd, modelo);
        return cmd;
    }

    protected override DbCommand ComandoEliminar(DbConnection conn, int id)
    {
        var cmd = CrearStoredProcedure(conn, "sp_DeleteServicio");
        CrearParametro(cmd, "p_id_servicio", id);
        return cmd;
    }

    protected override DbCommand ComandoLeer(DbConnection conn, int id)
    {
        var cmd = CrearStoredProcedure(conn, "sp_ListServicioById");
        CrearParametro(cmd, "p_id_servicio", id);
        return cmd;
    }

    protected override DbCommand ComandoLeerTodos(DbConnection conn)
    {
        return CrearStoredProcedure(conn, "spListServicios");
    }

    protected override Servicio MapearModelo(DbDataReader reader)
    {
        return new Servicio
        {
            idServicio = LeerEntero(reader, "id_servicio"),
            nombre = LeerTexto(reader, "nombre"),
            descripcion = LeerTexto(reader, "descripcion"),
            precioUSD = LeerDouble(reader, "precio_usd"),
            duracionHoras = LeerDouble(reader, "duracion_horas"),
            idiomaGuia = LeerTexto(reader, "idioma_guia"),
            capacidadMaxima = LeerEntero(reader, "capacidad_maxima"),
            incluyeRecojo = LeerBoolYN(reader, "incluye_recojo"),
            ciudadDestino = LeerTexto(reader, "ciudad_destino")
        };
    }

    private void SetCamposServicio(DbCommand cmd, Servicio modelo)
    {
        CrearParametro(cmd, "p_nombre", modelo.nombre);
        CrearParametro(cmd, "p_descripcion", modelo.descripcion);
        CrearParametro(cmd, "p_precio_usd", modelo.precioUSD);
        CrearParametro(cmd, "p_duracion_horas", modelo.duracionHoras);
        CrearParametro(cmd, "p_idioma_guia", modelo.idiomaGuia);
        CrearParametro(cmd, "p_capacidad_maxima", modelo.capacidadMaxima);
        CrearParametro(cmd, "p_incluye_recojo", modelo.incluyeRecojo ? "Y" : "N");
        CrearParametro(cmd, "p_ciudad_destino", modelo.ciudadDestino);
    }
}

File: Backend/ProyectoPro3Persistencia/src/main/java/pe/edu/pucp/proyectopro3/dao/reservas/ServicioDAOImpl.java
package pe.edu.pucp.proyectopro3.dao.reservas;

import pe.edu.pucp.proyectopro3.dao.DefaultBaseDAO;
import pe.edu.pucp.proyectopro3.modelo.reservas.Servicio;

import java.sql.*;

public class ServicioDAOImpl extends DefaultBaseDAO<Servicio> implements ServicioDAO {

    @Override
    protected PreparedStatement comandoCrear(Connection conn, Servicio modelo) throws SQLException {
        String sql = "{call sp_InsertServicio(?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);

        setCamposServicio(cmd, modelo, 1);
        cmd.registerOutParameter(9, Types.INTEGER);

        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, Servicio modelo) throws SQLException {
        String sql = "{call sp_UpdateServicio(?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);

        cmd.setInt(1, modelo.getIdServicio());
        setCamposServicio(cmd, modelo, 2);

        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, Integer id) throws SQLException {
        String sql = "{call sp_DeleteServicio(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, Integer id) throws SQLException {
        String sql = "{call sp_ListServicioById(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) throws SQLException {
        String sql = "{call spListServicios()}";
        return conn.prepareCall(sql);
    }

    @Override
    protected Integer extraerIdDesdeCallable(CallableStatement cmd) throws SQLException {
        return cmd.getInt(9);
    }

    @Override
    protected Integer extraerIdDesdeGeneratedKeys(ResultSet rs) throws SQLException {
        return rs.getInt(1);
    }

    @Override
    protected Servicio mapearModelo(ResultSet rs) throws SQLException {
        Servicio servicio = new Servicio();

        servicio.setIdServicio(rs.getInt("id_servicio"));
        servicio.setNombre(rs.getString("nombre"));
        servicio.setDescripcion(rs.getString("descripcion"));
        servicio.setPrecioUSD(rs.getDouble("precio_usd"));
        servicio.setDuracionHoras(rs.getDouble("duracion_horas"));
        servicio.setIdiomaGuia(rs.getString("idioma_guia"));
        servicio.setCapacidadMaxima(rs.getInt("capacidad_maxima"));
        servicio.setIncluyeRecojo("Y".equalsIgnoreCase(rs.getString("incluye_recojo")));
        servicio.setCiudadDestino(rs.getString("ciudad_destino"));

        return servicio;
    }

    private void setCamposServicio(CallableStatement cmd, Servicio modelo, int inicio) throws SQLException {
        cmd.setString(inicio, modelo.getNombre());
        cmd.setString(inicio + 1, modelo.getDescripcion());
        cmd.setDouble(inicio + 2, modelo.getPrecioUSD());
        cmd.setDouble(inicio + 3, modelo.getDuracionHoras());
        cmd.setString(inicio + 4, modelo.getIdiomaGuia());
        cmd.setInt(inicio + 5, modelo.getCapacidadMaxima());
        cmd.setString(inicio + 6, modelo.isIncluyeRecojo() ? "Y" : "N");
        cmd.setString(inicio + 7, modelo.getCiudadDestino());
    }
}
